src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+   it executes. Barbara adds runtime checks to identify this situation.
+1. `Waker` is `Sync`, so on a multithreaded system, a future could give another
+   thread access to its `Waker` and the other thread could wake it up. This
+   could happen while the `poll` is executing, before `poll` returns
+   `Poll::Pending`. Therefore, Barbara concludes that if `wake` is called while
+   a future is being polled then the future should be re-polled, even if the
+   current `poll` returns `Poll::Pending`. This requires putting a retry loop
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future finishes executing,
+   and the executor must ignore those wakeups. This duplicates the "ignore
+   spurious wakeups" functionality that exists in the future itself.
+
+Ultimately, this made the executor logic [quite
+nontrivial](https://github.com/tock/design-explorations/blob/master/size_comparison/futures/src/task.rs).
+The executor logic can be monomorphized for each future, which allows the
+compiler to make inlining optimizations, but results in a significant amount of
+duplicate code. Alternatively, it could be adapted to use function pointers or
+vtables to avoid the code duplication, but then the compiler *definitely* cannot
+inline `Future::poll` into the kernel callbacks.
+
+At this point, Barbara realizes she needs to do some benchmarking. Barbara comes
+up with a sample application -- an app that blinks a led and responds to button
+presses -- and implements it twice. One implementation does not use `Future` at
+all, the other does. In addition to two async drivers (a timer driver and a GPIO
+driver), the `Future`-based app uses Barbara's executor as well as a single
+combinator (that manages the state of the app).
+
+Barbara publishes an
+[analysis](https://github.com/tock/design-explorations/tree/master/size_comparison)
+of the relative sizes of the two app implementations, finding a large percentage
+increase in both code size and RAM usage (note: stack usage was not
+investigated). Barbara concludes that **`core::future::Future` is not suitable
+for highly-resource-constrained environments** due to the amount of code and RAM
+required to implement executors and combinators.
+
+Barbara redesigns the library she is building to use a [different
+concept](https://github.com/tock/design-explorations/tree/master/zst_pointer_async)
+for implementing async APIs in Rust that are much lighter weight. She has moved
+on from `Future` and is refining her [async
+traits](https://github.com/tock/libtock-rs/blob/master/core/platform/src/async_traits.rs)
+instead.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * `core::future::Future` isn't suitable for every asynchronous API in Rust.
+      `Future` has a lot of capabilities, such as the ability to spawn
+      dynamically-allocated futures, that are unnecessary in embedded system.
```

Comment:
Is it `Future` itself or rather the wake based API? Also, are we sure that the tradeoffs in code size experienced here are fundamental to getting `Future` to work? Or is it possible that there can be tweaks to the design that will allow these tradeoffs to go away?

Reply:
I didn't really distinguish `Future` from `Waker`.

I think there are fundamental tradeoffs at play. The fact that `Future` is demand-driven results in an information loss during callbacks. Kernel callbacks need to stash data somewhere for `poll` to retrieve later, and `poll` has to contain that retrieval logic. That ended up being the majority of the cost.